### PR TITLE
Add React/Jest testing infrastructure

### PR DESCRIPTION
**Pull request by trueadm**
Release note: none

This PR adds the React/Jest testing infrastructure to Prepack and is stacked on the PR https://github.com/facebook/prepack/pull/1117. Thus, that PR should be merged before this PR is merged to reduce noise in the diff.

This will allow us to test various runtime outputs of React 16 when running original source vs Prepacked source to see if there are any issues/differences that might have an impact on applications. The primary reason for this is to track regressions for the component folding PRs that will land in the future.

Please note, this PR does not contain any reconciler changes, thus `__registerReactComponentRoot(App);` has been commented out of tests to ensure they don't prematurely fail. A follow up PR will enable them once the other React functional component folding PRs get landed.

This PR also adds some mock React globals to be used within tests (and maybe to be further integrated into folding at a future point too). The mocks include `React.Component` and `React.cloneElement` for now.

Furthermore, a `utils/json.js` utility file exists to help normalize the results from the React test renderer so that adjacent JSON text nodes get merged, which is something that may exist because of how the reconciler (once the PR lands) handles inlining of child nodes.

The command to run React tests is `yarn test-react`.

### CODE REVIEW

File: scripts/test-react.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    throw new Error("React test runner issue");
+  }
+  // // Use the original version of the test in case transforming messes it up.
+  let { getTrials } = A;
+  // // Run tests that assert the rendered output matches.
+  let resultA = getTrials(rendererA, A);
+  let resultB = getTrials(rendererB, B);
+
+  // // the test has returned many values for us to check
+  if (Array.isArray(resultA) && Array.isArray(resultA[0])) {
```
Comment (gaearon) [inline scripts/test-react.js]: Can we restructure this so that instead of repeating similar logic in both branches, we first wrap `resultA` and `resultB` into an array if necessary and then do the assertions unconditionally?

File: scripts/test-react.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      await runTest(directory, "return-text.js");
+    });
+
+    it("Return undefined", async () => {
+      // this test will cause a React console.error to show
+      // we monkey patch it to stop it polluting the test output
+      // with a false-negative error
+      global.console.error = () => {};
+      await runTest(directory, "return-undefined.js");
+      global.console.error = originalConsoleError;
```
Comment (gaearon) [inline scripts/test-react.js]: Maybe `try { } finally { }` so this is isolated from other tests in case it fails?

File: src/intrinsics/react/global.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    writable: true,
+    enumerable: false,
+    configurable: true,
+  });
+  // require("SomeModule") support (makes them abstract)
+  let type = Value.getTypeFromName("function");
+  let requireValue = AbstractValue.createFromTemplate(
+    realm,
+    buildExpressionTemplate("require"),
+    (type: any), // Flow complains that type of Value isn't compatible with class Value :/
```
Comment (gaearon) [inline src/intrinsics/react/global.js]: Would be nice to understand why.

File: src/intrinsics/react/global.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/**
```
Comment (gaearon) [inline src/intrinsics/react/global.js]: It this only used for testing infra? Or for real implementation? It's not clear.
Should this be renamed to `src/intrinsics/react-mocks/global.js`?

File: src/utils/json.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+type JSONValue = Array<JSONValue> | string | number | JSON;
+type JSON = { [key: string]: JSONValue };
+
+// this will mutate the original JSON object
+export function normalize(node: JSON) {
```
Comment (gaearon) [inline src/utils/json.js]: This utility name sounds very generic but it's not very clear what this does.

What is normalized JSON looking like?
Is it useful in any other places than this test infra?

File: test/react/functional-components/clone-element.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+if (this.__createReactMock) {
+  var React = __createReactMock();
+} else {
+	var React = require('react');
```
Comment (gaearon) [inline test/react/functional-components/clone-element.js]: Nit: indent

File: test/react/functional-components/conditional.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+function MaybeShow(props) {
+  if (props.show) {
+    return props.children;
+  }
+  return null;
+}
+
+function App() {
+	return (
+		<MaybeShow show={true}>
```
Comment (gaearon) [inline test/react/functional-components/conditional.js]: Nit: indent

Comment (gaearon) [reply test/react/functional-components/conditional.js]: (and in most places below)

File: test/react/functional-components/dynamic-context.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// we can't use ES2015 classes in Prepack yet (they don't serialize)
+// so we have to use ES5 instead
+var StatefulComponent = (function (superclass) {
+  function StatefulComponent () {
+    superclass.apply(this, arguments);
+  }
+
+  if ( superclass ) StatefulComponent.__proto__ = superclass;
```
Comment (gaearon) [inline test/react/functional-components/dynamic-context.js]: Nit: style is odd, can we use the same code style as in code?

File: test/react/functional-components/dynamic-context.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+		return <Child />;
+	};
+	StatefulComponent.childContextTypes = {
+		title: () => {},
+	};
+
+  return StatefulComponent;
+}(React.Component));
+
+this.StatefulComponent = StatefulComponent;
```
Comment (gaearon) [inline test/react/functional-components/dynamic-context.js]: Why do we need `this.*` assignment here but not in other places?

File: scripts/test-react.js (hunk 2 above)
Comment (trueadm) [reply scripts/test-react.js]: I thought this wasn't actual JS error that is occurring, rather it's `console.error` being fired?

**Comment by sebmarkbage** [comment]
The move from `test/serializer/react/`→ `test/serializer/jsx/`. That implementation of JSX is still React specific. I've tried to keep the brand name around JSX separate from React. So it would be good if we could call this something with React in it. `test/serializer/react/jsx/` maybe?

File: src/intrinsics/prepack/global.js [hunk 10]
```diff
@@ -155,6 +155,39 @@ export default function(realm: Realm): void {
     configurable: true,
   });
 
+  if (realm.react.enabled) {
+    global.$DefineOwnProperty("__reactComponentRoots", {
```
Comment (sebmarkbage) [inline src/intrinsics/prepack/global.js]: @cblappert This is unrelated to this PR. I saw that this pattern was added in the additional functions one, but why do we make these a public object instead of just an internal Map? Seems weird for the compiled script to be able to reason about this and mutate this.

File: scripts/test-react.js (hunk 2 above)
Comment (gaearon) [reply scripts/test-react.js]: If `runTest` fails for some reason, `global.console.error` won't be restored.

If you do

```js
global.console.error = ...
try {
  doStuff
} finally {
  global.console.error = original
}
```

it is restored no matter what.
